tests/gradients/core/test_pulse_generator_gradient.py

```diff
@@ -0,0 +1,1123 @@
+    )
+    def test_single_zeros(self, num_wires, remove_ids):
+        """Test that a removing single entries/Pauli words works as expected."""
+        dim = 4**num_wires - 1
+        # Set coefficients to zero
+        coeffs = tuple(np.zeros(dim) if i in remove_ids else e for i, e in enumerate(np.eye(dim)))
+        new_coeffs, words = _nonzero_coeffs_and_words(coeffs, num_wires)
+
+        mask = np.ones(dim, bool)
+        mask[remove_ids] = 0
+        exp_coeffs = tuple(c[mask] for c in coeffs)
+        exp_words = [w for i, w in zip(mask, pauli_basis_strings(num_wires)) if i]
+        assert all(qml.math.allclose(nc, c) for nc, c in zip(new_coeffs, exp_coeffs))
+        assert len(words) == 4**num_wires - 1 - len(remove_ids)
+        assert all(w == exp for w, exp in zip(words, exp_words))
+
+        # Remove coefficients altogether
+        coeffs = tuple(e for i, e in enumerate(np.eye(dim)) if i not in remove_ids)
+        new_coeffs, words = _nonzero_coeffs_and_words(coeffs, num_wires)
+        exp_coeffs = tuple(c[mask] for c in coeffs)
+        assert all(qml.math.allclose(nc, c) for nc, c in zip(new_coeffs, exp_coeffs))
+        assert len(words) == 4**num_wires - 1 - len(remove_ids)
+        assert all(w == exp for w, exp in zip(words, exp_words))
+
+    def test_atol(self):
+        """Test that the precision keyword argument atol is used correctly."""
+        atols = [1e-8, 1e-4, 1e-1]
+        coeffs = (np.array(atols), np.array(atols) * 2)
+        for i, atol in enumerate(atols):
+            new_coeffs, words = _nonzero_coeffs_and_words(coeffs, 1, atol=atol / 10)
+            assert len(new_coeffs) == 2
+            assert all(len(c) == 3 - i for c in new_coeffs)
+            assert words == ["X", "Y", "Z"][i:]
+        new_coeffs, words = _nonzero_coeffs_and_words(coeffs, 1, atol=1.0)
+        assert new_coeffs == [[], []]
+        assert words == []
+
+
+tapes = [
+    ([qml.PauliX(0)], []),
+    (["evolve_op"], [qml.expval(qml.PauliZ(0))]),
+    (["evolve_op", qml.RZ(0.3, "b"), "evolve_op", qml.PauliX(0)], []),
+]
+
+tapes_and_op_ids = [
+    (tapes[0], 0),
+    (tapes[0], 1),
+    (tapes[1], 0),
+    (tapes[1], 1),
+    (tapes[2], 0),
+    (tapes[2], 3),
+    (tapes[2], 4),
+]
+all_ops = [[rot := qml.PauliRot(0.3, "IXZ", [0, 1, "a"])], [rot, qml.RY(0.3, 5), qml.PauliX(1)]]
+
+
+@pytest.mark.jax
+class TestInsertOp:
+    """Test the utility _insert_op."""
+
+    # pylint: disable=too-few-public-methods
+
+    @pytest.mark.parametrize("tape, op_idx", tapes_and_op_ids)
+    @pytest.mark.parametrize("ops", all_ops)
+    def test_output_properties(self, tape, op_idx, ops):
+        """Test that the input tape and inserted ops are taken into account correctly."""
+        evolve_op = qml.evolve(qml.pulse.constant * qml.PauliZ("a"))([np.array(0.2)], 0.2)
+        operations, measurements = tape
+        operations = [evolve_op if op == "evolve_op" else op for op in operations]
```

Comment:
Why first the abstraction with the place-holder string and then the constant `evolve_op` inside the test? Could you not create `evolve_op` outside the test and already insert it in the operations parametrization? Am I missing something?

Reply:
Because `ParametrizedEvolution` needs `JAX`, and creating the op outside (but within the test class) does not work, even when using `@pytest.mark.jax` :/

Reply:
🙈 aaaaah